python/tvm/topi/cuda/nms.py

```diff
@@ -97,47 +97,44 @@ def get_valid_counts_ir(
+    if isinstance(score_threshold, float):
+        score_threshold = tvm.ir.make_node("FloatImm", dtype="float32", value=score_threshold)
     id_index = tvm.ir.make_node("IntImm", dtype="int32", value=id_index)
     score_index = tvm.ir.make_node("IntImm", dtype="int32", value=score_index)
 
     max_threads = int(tvm.target.Target.current(allow_none=False).max_num_threads)
-    nthread_tx = max_threads
-    nthread_bx = batch_size * num_anchors // max_threads + 1
-    tx = te.thread_axis("threadIdx.x")
-    bx = te.thread_axis("blockIdx.x")
-    ib.scope_attr(tx, "thread_extent", nthread_tx)
-    ib.scope_attr(bx, "thread_extent", nthread_bx)
-    tid = bx * max_threads + tx
-    idxd = tvm.tir.indexdiv
-
-    # initialize valid_count
-    with ib.if_scope(tid < batch_size):
-        valid_count[tid] = 0
-    with ib.if_scope(tid < batch_size * num_anchors):
-        i = idxd(tid, num_anchors)
-        with ib.if_scope(
-            tvm.tir.all(
-                data[tid * elem_length + score_index] > score_threshold,
-                tvm.tir.any(id_index < 0, data[tid * elem_length + id_index] >= 0),
-            )
-        ):
-            atomic_add_return[0] = atomic_add(
-                tvm.tir.call_intrin("handle", "tir.address_of", valid_count[i]), one_count
-            )
-            with ib.for_range(0, elem_length) as k:
-                out[tid * elem_length + k] = data[tid * elem_length + k]
-                out_indices[tid + k] = tid + k
-        with ib.else_scope():
-            with ib.for_range(0, elem_length) as k:
-                out[tid * elem_length + k] = -one
-                out_indices[tid + k] = -one_count
-
+    with ib.new_scope():
+        nthread_tx = max_threads
+        nthread_bx = batch_size // max_threads + 1
+        tx = te.thread_axis("threadIdx.x")
+        bx = te.thread_axis("blockIdx.x")
+        ib.scope_attr(tx, "thread_extent", nthread_tx)
+        ib.scope_attr(bx, "thread_extent", nthread_bx)
+        tid = bx * max_threads + tx
+        with ib.if_scope(tid < batch_size):
+            valid_count[tid] = 0
+            i = tid
+            with ib.for_range(0, num_anchors) as j:
+                score = data[(i * num_anchors + j) * elem_length + score_index]
+                with ib.if_scope(
+                    tvm.tir.all(
+                        score > score_threshold,
+                        tvm.tir.any(
+                            id_index < 0, data[(i * num_anchors + j) * elem_length + id_index] >= 0
+                        ),
+                    )
+                ):
+                    with ib.for_range(0, elem_length) as k:
+                        out[(i * num_anchors + valid_count[i]) * elem_length + k] = data[
+                            (i * num_anchors + j) * elem_length + k
+                        ]
+                    out_indices[i * num_anchors + valid_count[i]] = j
+                    valid_count[i] += 1
```

Comment:
There's data race between line 129 and line 133 for `valid_count` for loading and writing.

Reply:
There is definitely not a data race now, because I removed the threading :smile: But I think I see your point, this might be while I couldn't pass the test with threading on. I will investigate

Reply:
Could you use atomic add here?

Reply:
atomic_add doesn't work with nvptx. That's a headache...

Reply:
I do not have a race condition here because I'm not parallelizing over that axis. The issue with the current kernel is that it cannot implement the ONNX API because it cannot properly sort the output indices, and ONNX needs those outputs in the correct order to do some post-processing.

Reply:
I see. So we lose parallelism in `num_anchors` compared with the original implementation. Are we able to keep the level of parallelism while getting the correct output indices?

Reply:
The test is disabled on master:
https://github.com/apache/tvm/blob/7a0d10ce8d4df27734949b944992e5c516f620d0/tests/python/relay/test_op_level5.py#L317-L321

The current kernel doesn't follow the spec, specifically, it uses the wrong index for the out_indices update:
```
out_indices[tid + k] = tid + k
```
instead of
```
out_indices[i * num_anchors + valid_count[i]] = j
```
Unfortunately, because valid_count depends on the iteration order of the loop on this axis, I don't see a way to get the correct index without serializing this axis.

Reply:
If you have a way to get the right behavior with more threading, I'm very interested in hearing it.

Reply:
To have maximum parallelism, we can do two iterations with the first marking all valid elements with flag 1, and the second doing a prefix sum to get the correct output index.

Reply:
I think it takes three iterations and four extra output-allocated tensors

The first is what's currently in main, but outputting to a tmp_out and a tmp_out_indices, plus an is_valid tensor.

The second is a kernel to do cumsum on is_valid to get sort_indices

The third is to re-order the outputs based on the sort_indices.

It's a little messy. I'm reading gpugems on parallel cumsum, but I'll give it a try.

Reply:
Yes, that method was actually my first implementation of nms before PR #5339. And in that PR, what I improved was only getting the correct valid_count from nms_ir, and marking -1 to invalid outputs. And after nms_ir, I do a sort (in descending order), to get the sorted outputs as well as sorted output indicies. The second implementation is currently in main and much faster than the previous one.

Reply:
So for your case, if you want to pass the correct sorted output and output indices after `get_valid_counts_ir`, what you need to do is to do an argsort with the value `out` and indices `out_indices` which you get from `get_valid_counts_ir` from current main: 
```sort_tensor = argsort_thrust(
            score_tensor, valid_count=None, axis=1, is_ascend=False, dtype=valid_count_dtype)
```
And that is what did at the beginning of nms: https://github.com/apache/tvm/blob/main/python/tvm/topi/cuda/nms.py#L561

I don't pass in `valid_count` because I haven't reordered the invalid elements to the end of the output array yet but only mark those invalid elements as -1. But after sorting the whole array, the invalid elements will be sorted to the end of the array. So will their indices.

Reply:
I don't see the sort?

It's faster, but it's not correct, that PR also disabled the relevant tests I need to ensure ONNX works properly.

Reply:
I'll go back and look at what you had.

Reply:
Do you happen to have a script to reproduce the benchmarks you had in that PR? I assume it was a TF import?

Reply:
It was actually a mxnet model and the model is a customized detection model, and `num_anchors` is really large. You can run this tutorial for a benchmark reference: https://tvm.apache.org/docs/tutorials/frontend/deploy_ssd_gluoncv.html

Reply:
Actually, I mentioned in PR #5339 that the model is `ssd_resnet50_v1` with (1, 3, 512, 512) input shape and with Thrust turned on in build